### PR TITLE
Add global comeback feature

### PR DESCRIPTION
**Pull request by yammesicka**
Solves #17 

Also: add `.vscode/` to `.gitignore`

### CODE REVIEW

File: comeback/main.py [hunk 1]
```diff
@@ -110,6 +118,16 @@ def load_config() -> None:
     run_config(config)
 
 
+def create_comeback_file_here() -> pathlib.Path:
+    verbose_echo('Creating a blank .comeback configuration file.')
+    path = paths.CURRENT_DIR / '.comeback'
+    if path.exists():
+        print('There is already .comeback file here. Will only touch it.')
```
Comment (agamm) [inline comeback/main.py]: why not `verbose_echo`?

File: comeback/config.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                         -> Iterator[Tuple[str, int]]:
+    comeback_files = get_comeback_paths()
+    date_getter = operator.itemgetter(1)
+    sorted_ = sorted(comeback_files.items(), key=date_getter, reverse=True)
+    yield from itertools.islice(sorted_, max_results)
+
+
+def get_last_comeback() -> pathlib.Path:
+    FILEPATH_INDEX = 0
+    single_comeback = get_recent_comebacks(max_results=1)
```
Comment (agamm) [inline comeback/config.py:62]: why only 1? :(

Comment (yammesicka) [reply comeback/config.py:62]: Because it's `*last*_comeback` :P

Comment (agamm) [reply comeback/config.py:62]: I mean why only show one? why not a bunch like 5 last ones? (more of a practical question).
I maybe want to go to some older project this time.

I get it that you are loading it directly, but I think we should present a numbered list and give an `input` for the user to choose which one, no?

File: comeback/config.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+
+from comeback import paths
+
+
+EPOCH = Union[int, float]
+
+
+"""
+This code is for more complicated cases where we may need to save more data.
```
Comment (agamm) [inline comeback/config.py]: move to a different branch?

File: comeback/config.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    except json.JSONDecodeError:
+        return create_and_get_basic_comeback_file()
+
+
+def write_comeback_paths(paths_to_write: Dict[str, int]) -> None:
+    paths_file = paths.PATHS_DATA_FILE
+    paths_file.write_text(json.dumps(paths_to_write))
+
+
+def add_comeback_path(path: Optional[Union[pathlib.Path, str]] = None,
```
Comment (agamm) [inline comeback/config.py]: When will we ever define `last_used`?

File: comeback/config.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                      last_used: Optional[int] = None) -> None:
+    path = path if path is not None else pathlib.Path.cwd() / '.comeback'
+    if not pathlib.Path(path).exists():
+        raise FileNotFoundError(f'Can\'t locate {path}')
+    path_to_add = create_path_entity(filepath=path, last_used=last_used)
+    paths_to_write = {**get_comeback_paths(), **path_to_add}
```
Comment (agamm) [inline comeback/config.py:48]: Nice append :)

File: comeback/config.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    write_comeback_paths(paths_to_write)
+
+
+def get_recent_comebacks(max_results: Optional[int] = None)\
+                         -> Iterator[Tuple[str, int]]:
+    comeback_files = get_comeback_paths()
+    date_getter = operator.itemgetter(1)
+    sorted_ = sorted(comeback_files.items(), key=date_getter, reverse=True)
+    yield from itertools.islice(sorted_, max_results)
```
Comment (agamm) [inline comeback/config.py:57]: noice

File: comeback/main.py [hunk 7]
```diff
@@ -22,7 +23,7 @@ def verbose_echo(msg: str) -> None:
 
 
 def get_probable_project_name() -> str:
-    return get_cwd().name
+    return paths.CURRENT_DIR.name
```
Comment (agamm) [inline comeback/main.py:26]: wow such knowledge of pathlib superbs

File: comeback/main.py [hunk 8]
```diff
@@ -99,9 +100,16 @@ def read_config_file(config_path: pathlib.Path) -> Optional[Dict[str, Any]]:
...
+    if cwd_path.exists():
+        return cwd_path
+    return config.get_last_comeback()
+
+
 def load_config() -> None:
-    verbose_echo(f'Loading configuration file form: {get_cwd()}')
-    config_path = get_cwd() / '.comeback'
+    verbose_echo(f'Loading configuration file form: {paths.CURRENT_DIR}')
+    config_path = get_config_path()
```
Comment (agamm) [inline comeback/main.py:112]: such DRY, good good

File: comeback/paths.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+def get_default_comeback_file_path() -> pathlib.Path:
+    return pathlib.Path.home() / '.comeback'
+
+
+
+DEFAULT_COMEBACK_FILE = get_default_comeback_file_path()
+DATA_DIR = get_data_path()
+CONFIG_DIR = get_config_path()
+PATHS_DATA_FILE = get_data_path(filename='paths')
+CURRENT_DIR = pathlib.Path.cwd()
```
Comment (agamm) [inline comeback/paths.py:53]: cheater! I take back my knowledge comment 💢

Comment (agamm) [reply comeback/paths.py:53]: Nice separation though.

File: comeback/paths.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class XdgPath(NamedTuple):
+    name: str
+    xdg_variable_name: str
+    default_dir: pathlib.Path
+
+
+HOMEDIR = pathlib.Path.home()
+PATH_KINDS = {
```
Comment (agamm) [inline comeback/paths.py:16]: Maybe add a comment with a URL to explain this?
Like [https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html](https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html)

Comment (agamm) [reply comeback/paths.py:16]: For example, I was really confused as to why you used `config` and `data` terminology.

File: comeback/config.py (hunk 2 above)
Comment (yammesicka) [reply comeback/config.py:62]: It's just a specific case. You can use `get_recent_comebacks` with or without parameters directly if you want to fetch more.

I think that the user shouldn't be bothered with prompts when he just tries `comeback`. We can implement `comeback --list` that will show all the recent entries and will allow the user to choose. :)

File: comeback/paths.py (hunk 9 above)
Comment (yammesicka) [reply comeback/paths.py:53]: XDXD

File: comeback/paths.py (hunk 10 above)
Comment (yammesicka) [reply comeback/paths.py:16]: Sure, why not

File: comeback/config.py (hunk 4 above)
Comment (yammesicka) [reply comeback/config.py]: Maybe we would like to hide some entry in the future(?)

File: comeback/config.py (hunk 3 above)
Comment (yammesicka) [reply comeback/config.py]: Kk

File: comeback/config.py (hunk 2 above)
Comment (agamm) [reply comeback/config.py:62]: Ok, for now, it's good.

File: comeback/config.py (hunk 4 above)
Comment (agamm) [reply comeback/config.py]: Better to use a meta key of `hide`.

**Comment by agamm** [comment]
@yammesicka ping me when I should approve after changes ;)

**Review by agamm — APPROVED** [review]
Stamp of approval granted. 
http://gph.is/2rMkoJZ

**Pull request merged**